analysis/src/vm/function_desugar.rs

```diff
@@ -0,0 +1,207 @@
+//! Desugar asm functions which operate on values to functions which operate on witness columns
+//! TODO: Using the same term "function" for both is a bit awkward
+
+use ast::{
+    asm_analysis::{
+        utils::previsit_expression_mut, AnalysisASMFile, AssignmentStatement, FunctionStatement,
+        FunctionStatements, InstructionStatement, Machine,
+    },
+    parsed::{
+        asm::{Param, ParamList},
+        Expression,
+    },
+};
+use number::FieldElement;
+use std::{collections::HashMap, marker::PhantomData, ops::ControlFlow};
+
+use crate::utils::{parse_instruction_definition, parse_register_declaration};
+
+/// Desugar VM functions
+pub fn desugar<T: FieldElement>(file: AnalysisASMFile<T>) -> AnalysisASMFile<T> {
+    FunctionDesugar::default().generate(file)
+}
+
+#[derive(Default)]
+struct FunctionDesugar<T> {
+    marker: PhantomData<T>,
+}
+
+fn input_at(i: usize) -> String {
+    format!("_input_{}", i)
+}
+fn output_at(i: usize) -> String {
+    format!("_output_{}", i)
+}
+
+fn substitute<T>(mut e: Expression<T>, inputs: &HashMap<String, usize>) -> Expression<T> {
+    previsit_expression_mut(&mut e, &mut |e| {
```

Comment:
is there a version of this but for statements that also applies it to expressions? if yes we could maybe avoid the big replacement match lower in this file